### ISSUE TITLE
20x256 or 30x256

### ISSUE THREAD

**Issue by alphaladder** [gcp/leela-zero#1553]
15b is almost ready to its summit and should we start our new N2N from30x256?
 1) there is no big difference between 20b and 30b in the LZ play level. And 30b can easily exceed the elf weights while 20b couldn’t.
2) do not worry about the amount of self play games .in fact we have 30000 self play in 10b stage while we still have the same number in 15b stage. why? Big block means better Lz level and would be attractive for person on training .

**Comment by grolich** [comment]
There were a few speed optimizations in the process which allowed the relatively smooth transition to 15b.
The code became faster, countering the slowdown.


I would think transitioning to your suggestion of 30x256 would be a massive slowdown, for an improvement we are not sure about (how much more are we expecting out of the net vs. the time it will take and the risk of lack of progress scaring people away), but I may be wrong. 

As to exactly (or approximately) how much of a slowdown can be expected as the number of blocks gets higher without further speed optimizations is something @gcp or roy (or some of the other developers) are more capable of answering than me.

**Comment by PhilipFRipper** [comment]
Adding blocks is roughly linear, so 15 to 30 would be 1/2 the speed.  But going from 192 to 256 filters, I have no idea.

**Comment by john45678** [comment]
Isn't a goal of the project to have a net of each size? Also I'm sure most people want to know if 20b LZ can match 20b ELF.

**Comment by Marcin1960** [comment]
@john45678 "I'm sure most people want to know if 20b LZ can match 20b ELF."

I don't. Actually I do not care about ELF.

**Comment by alphaladder** [comment]
@grolich  based on our experience, when we upgrade the blocks, more person would like to upgrade their pc in order to unveil the mask of AGZ 40b. The more closer to 40b, the more attractive to potential trainer to join us. If we only stuck by the speed ( it varies on the pc and clients), the LZ is just a daydream. I heard the similars voice when we goes up from 10 to 15, now you saw only less than 2 months it reach the summit. Thus, do not worry about the speed. Otherwise, the trainers will lose their patience due to the no promotion of Lz.

**Comment by alphaladder** [comment]
By using 30b ,we could digest elf stick completely
 and advance to AGZ 40b steadily.

**Comment by gcp** [comment]
>15b is almost ready to its summit

What's the evidence for this? The 20b nets that have been trained were no better than the 15b ones.

**Comment by Crazyeight101** [comment]
So in lieu of someone magically providing a larger network that's stronger than the current 15b, what are the next steps?  Is there some cut-off in mind for training the 15b?  Are we going to experiment with using more than 250k ELF games before we do?

Even if the 15b hasn't plateaued yet, it certainly seems like we're getting close to that point, doesn't it?  The number of training games between promotions is suddenly growing quite quickly.

**Comment by PhilipFRipper** [comment]
@Crazyeight101 if it works the way 128x10 worked you can expect to hit 200,000+ games per engine at some points

**Comment by barrtgt** [comment]
Has anyone done a resign analysis after the elf adoption and test elf and lz nets separately?

**Comment by Crazyeight101** [comment]
@PhilipFRipper I'm aware there's still a little ways to go before the network is saturated, and with 250k ELF games included in the window, the real plateau will be somewhere between 250k-300k games between promotions, but if we continue going at this rate, it'll probably only be a handful of promotions  more, in all likelihood, before we reach that point.  The point is it's already a good time to consider what comes next, and I'm curious if any decisions have already been made.

**Comment by Friday9i** [comment]
Efficiency should drive the choice...
Up to now, the 15x192 nets were progressing fast, so it seemed adequate to pursue the efforts with this net size, which has the advantage to generate selfplay games quite quickly. But for ~3 weeks, things have slowed down a lot, and improvements are limited. What to do?
**Choice 1, we go on with 15x192:** even if we stay with it for a few months, we'll probably get at most 3 to 5 new nets, and it seems we will be capped around 11600 or 11700 selfplay Elo, which is a limited improvement vs now. Then we will switch to larger networks and improvements will be fast
**Choice 2, we go to larger networks very soon:** improvements will be quick, we will get much stronger networks in a few weeks and selfplay games will be much better. Hence, by the end of July, we will be way above today
What is the best choice? Choice 1 allows to better understand the limits of the 15x192 nets, but apart from that, it seems painful to spend months for limited improvements, while we could get important improvements fast while switching to larger networks.
**Hence, it seems a rapid switch to larger networks would be more efficient.** Should it be 20x256 or 30x256? No idea...

**Comment by Marcin1960** [comment]
I have another proposal.

Train few different sizes to see where the  gain/price sweet spots are. 20x192, 15x256, 30x128, etc ...

If not now, then when?

**Comment by jkiliani** [comment]
@gcp Is there a reason @bjiyxo's most recent 20x256 net was not queued for an (official) match? According to the unofficial match results in https://github.com/gcp/leela-zero/issues/1113, V15 should be at least slightly stronger than net 148, so it (or the successor net) should be a bootstrap candidate if there is an actual stall...

**Comment by afalturki** [comment]
I think regardless of the size of the bigger network. The size upgrade should be accompanied with some type of visits lowering. 256x20 at 800 visits would be around 2 times faster than current games/day.

**Comment by tapsika** [comment]
Lowering visits will likely make the stalling problem worse (after the initial jumps with the deeper net). It is unclear if LZ's proneness to stagnate is because of the selfplay labeling quality (poor training data) or because of some kind of capacity limit or obstacle in the NN. The strength difference to ELF (with not much bigger net) suggests that real capacity limits should be a bit farther though.

**Comment by jkiliani** [comment]
Actually no. Lowering visits **right now** may make the stalling worse, but noone suggested that. The time to lower visits is when changing network sizes, not before. It is completely logical that a 192x15 network cannot catch up to a 224x20 net that's maxed, so a stall like may be happening now would be almost certainly capacity related.

**Comment by Ishinoshita** [comment]
I have no idea if what I will propose now is acceptable in terms of training workload, but it has the potential to reduce the number of posts here....
Why not continuously train 3 networks in parallel on the selfplay games? One would be equivalent of current best network, say current 15x192, used for generating selfplay games. The second one would be the previous size network, say the 12x128, which training would have been continued with all the selfplay games generated by the  successive 15x192. The third one would be next candidate in size, say 20x256 e. g., that would have been implemented when at the time of jump to to 15x192 and trained on all data ever since.
Match would be pushed from time to time to see how they rank respect to each over. 
If current network (15x192) flattens but the gap with previous network does not widen, then the blame for plateau in cannot be put on network capacity. If next network does not plateaus in the same way as current one, poor training data issue can be ruled out. Etc.... That would also allow to assess at any time the opportunity of to move up in size. Would this happen, then current would be made previous, next be made current and a new bigger net started as next candidate, pre-trained with past data then on the run as the other two.

**Comment by tapsika** [comment]
> "Not much bigger net" doesn't make any sense, ELF is considerably bigger than 192x15, so why should it not be considerably stronger?

Total cognitive capacity for ELF is less than double (1.8), and recent LZ results with 20 blocks were not that impressive, so there is some possibility that ELF advantage may not be solely because of its few extra layers. If there are other differences (in training method / parameters for example), then 15 blocks may also be capable of more than this.

@Ishinoshita: If you think about such radical changes, then it would also make sense to keep using a pool of the last few nets for selfplay. More diversity, and less chance for forgetting. You only need to define a threshold for dropping a net from the pool, when it's search+net is not stronger enough to effectively teach the new net.

But we still have zero data on how much better should the search and its labeling be for making an effective teacher - which is why all the talks about lowering/increasing visit counts seem unfounded.

**Comment by wonderingabout** [comment]
i dont think 192x15 is at its limit.
it just started to go slower for some time, but if we wait new networks will be promoted like usual, and maybe at a faster rate.

With more GPU power, it can certainly improve with many more network promotions.
I think lz should keep 192x15 for a few more weeks and see how things go.

**Comment by afalturki** [comment]
It is possible that lowering visits could make stalling worse. However, that might not be true as setting visits at 800 would quadruple the number of games/day at the same time. So, it would take more new games to produce a new best network but still the same or even less time.

**Comment by Crazyeight101** [comment]
@afalturki the training window is limited for the net anyways, and we're likely well passed the point where quantity would be more valuable than quality.  Increasing the visits would make much more sense than decreasing the visits, as the games produced for training would be stronger and certainly lead to better weights and biases with the only issue being how cost of computing time.  But as stated above, there's currently no reason to change the visits from 3200, unless someone were to produce very compelling data and results advocating the increase/decrease.

@Ishinoshita I think that's a brilliant idea about training additional networks one size above and below the currently used one, as long as it's relatively easy to implement and doesn't require too much computing resources.

**Comment by alreadydone** [comment]
[This speed-up](https://github.com/gcp/leela-zero/pull/1497) has been merged into /next.
Data for lowering visits at https://github.com/gcp/leela-zero/issues/1416.

**Comment by wonderingabout** [comment]
@Crazyeight101 
i understand your reasoning, sooner or later lz will have to go to 20 blocks, but i think lz 147 and 148 were just stronger than a simple promotion. They both got arround 57-58% winrate, so you could say its as if they got 2 promotions on a row. Some strongers networks than lz 148 already appared, but its not significant because <55%.
I believe the early elf boost was a bit unrealistic, it was like you said a temporary boost that had less impact over time. So i think the natural course for a 58% network is to take more time to be beaten by a significantly stronger (>55.0% network), if not for the early elf boost.

So i think if we just wait a few more days, its possible that a slightly stronger network appears 55.0%, which would be beaten sooner than a 58% network.
Maybe i'm wrong in all this, and if yes feel free to correct me.

**Comment by l1t1** [comment]
@wonderingabout the 0.58 self-player games' quaility also higer than 0.55, why they did not speed up promote?

**Comment by nutpen85** [comment]
Just an idea, which I don't like very much but maybe some of you do: If we want more steady progress, one could lower the 55% winrate while increasing the number of match games, thereby maintaining the significance. However, I'm more the patient type, who likes to wait for the next promotion under current rules.

**Comment by Marcin1960** [comment]
@nutpen85: "If we want more steady progress, one could lower the 55% winrate"

It might be that this supposed flattening is not a sign of the net exhausting its potential, but of a need for smaller steps like 52% instead of 55% at the pressent condition.

To illustrate: let us assume that this percentage number is always precise and non random. Would you think that it is strange when advances were decreasing, from 56%, 54% ,57%,49%, 58% to 54%, 55%, 51%, 48% , 52%?

Or like an alpinist as higher he gets, each step upward gets smaller on average, long before he reaches summit? He should not give up, just because the steps are not big enough. The standard he followed at the low altitude has to be relaxed.

I think, it is time to reduce the requirement, to 52% or 51% to allow the net grow in its own pace instead of waiting for the spectacular jump possible at lower level.

**Comment by wonderingabout** [comment]
i think 55% is fine, it avoids false positives and really there's no rush to improve fast. 
lz was not made to beat alphago, but at a collaborative and experimental public project, what matters is not the raw strength but how to achieve it.
Personally, i find interesting to see how things develop at this pace, and if in the next weeks the trend is the same then possibly go to 20 blocks, but i think it's a bit too soon now, we need to observe more

**Comment by Marcin1960** [comment]
"i think 55% is fine, it avoids false positives"

Well, because there are random fluctuations, if waiting long enough ("a few more weeks") some weaker nets WILL pass the 55% barrier. 

According to your approach 57% would be even better.
But what if the real advances are around 52% now?

**Comment by wonderingabout** [comment]
"Well, because there are random fluctuations, if waiting long enough ("a few more weeks") some weaker nets WILL pass the 55% barrier."

yes, but it doesnt matter, because statistically much more real stronger ones will pass.
however, if you decrease the gating to 52% more false positives will pass, leading to possible less progression or worse regressions in strength in the long run

**Comment by Marcin1960** [comment]
@wonderingabout "you decrease the gating to 52% more false positives will pass, leading to possible less progression or worse regressions in strength in the long run"

But what if at present actual advances are in the 51%-54% range? And passing 55% will require a random fluke?

If so, it would mean rejecting many actual advances to pick one randomly one of them in order to avoid one weak one, that would be eliminated quickly anyway.

**Comment by wonderingabout** [comment]
@Marcin1960 
it will advance eventually, given enough time and computing power.
Like i said earlier, one month of waiting with 1 computing power unit equals 1 day with 720 computing power units.
Since we dont have this computing power, we can only wait.
It will be the same with 20 blocks too, but the games will be longer to generate, making it even longer to wait, so i think its more comfortable to wait a few more weeks on 15 blocks while we can

**Comment by wonderingabout** [comment]
@Friday9i 
did deepmind compare 50% to 55% gating and say that 50% is superior ?
it's possible that 55% achieves the same as 50%, if not better.

also, for a big project like alphago, the 2 of 52% is significant because they have much much more computing power.

but for a small project with limited computing power like lz, if it starts to loose strength then you'll have to wait a lot more for it to get its strength back, not to mention improving.
So actually, i believe a higher gating is safer and maybe more efficient for a small project like lz

**Comment by Friday9i** [comment]
Deepmind began with 55%, then they switched to 45% (ie no-gating) and demonstrated it can work (at least in the context they created). They did not compare 50% vs 55% gating to my knowledge, and I think that the no-gating test was to demonstrate "it can work", not sure at all it's the most efficient. But it's nicer and more remarkable to say "the unexpected no-gating works" than to say "after testing, it seems that 51% (or whatever) gating is the most efficient". So to me, unless we get more info from them, it's a proof of concept, not the proof it's the most efficient (and it will depend on context in any case).

**Comment by wonderingabout** [comment]
@Friday9i 

fine then, if you want to lower the gating then you'll have to gain more significance.
you saw how during a 400 game match the winrate can fluctuate between 45 then go up to 53 then redrop to 49.

so to lower the gating to 51%, you'd have to play 1500-2000 games per match.
is it really efficient and worth it ?
isnt the current 400 games/55% system more efficient, but also a lot less hassle free for the community (no strength drops)

**Comment by Friday9i** [comment]
Exactly what I took into account in #1524!

**Comment by Friday9i** [comment]
It works well with 400 match games and 50% (or 52%) gating, no need to make more match games :-)

**Comment by jkiliani** [comment]
That's totally unproven, the argumentation in #1524 is flawed by not taking into account the effects on training, in particular by making the starting point reset less effective.

**Comment by Friday9i** [comment]
Unproven too and unprovable (the biais I spoke about from the beginning), hence we won't get a proof, hence we will never try if I follow you...

**Comment by wonderingabout** [comment]
really, i really think the 55% gating is, like i explained in my above messages, so much hassle free for a community project.

No strength drops, slow but steady improvements, no worry (and people worry very easily)
and for a community with limited computing power, its much more reliable and hassle free (community will easily worry about strength drops, that can last for a long time because of relatively low computing power of the community)

i dont see why you want so much to lower the gating. ok, it will make things more dynamic, but more dynamism equals more trouble...
and again, its not proven that 50% will bring stronger results than 55%, and even if it were, i'd still rather stick with the slow but reliable 55% gating

**Comment by PhilipFRipper** [comment]
The bloat of comments is detrimental to the project, making it harder for the devs to see any important posts.

**Comment by Friday9i** [comment]
Proof: it's unprovable mathematically speaking, as changing the selection process impacts the distribution of strength of new networks in an unknown way.
"No way  to go faster": yes there is one, it is to be more efficient! Is 50% gating more efficient? Not sure, but it seems 50% or 52% may be more efficient, for various credible distribution of strength (play with the Excel, it's hard to find distribution were it is less efficient AND realistic! Not impossible, but hard!).
"Risky": not so much! I did the effort to test it: try it by yourself and get a feeling of how it works before saying "what if it decrease strength for 10 days"! While 45% threshold (no gating from AZ) is risky, as seen with Minigo and LC0, il can obviously work (AZ). In between, around 50% threshold, it's hard to find circumstances where it decreases the level of play, so the risk is quite low, and at 52% it seems very low (and more efficient, once again).
"Just add more computing power": my GTX1080 already works on it... But gaining a few % efficiency is 'several GTX1080Ti', more than what I can put reasonably on the project ; -), and it may be 10% additional efficiency through better gating. Not sure once again, but it could be (and cannot be proved, so if we don't try we won't know)
Eventually, lower gating will give more diversity in networks, which could help by giving more diversity in selfplay games, that could be a positive and additional side effect.

**Comment by Crazyeight101** [comment]
@Friday9i @wonderingabout @Marcin1960 Agree with jkiliani and Philip.  Please don't create 15+ comments continuously rehashing arguments that have already been made.

At this point, we've already produced over 250k games since the last promotion, with 190k coming from d018.  Clearly we've run into an obstacle, but we don't actually know what it is and we're not taking steps to actually identify the problem or disprove our theories about it.  Cases have been made for lowering gating, changing visits, removing ELF games, moving to a larger network, and so on, but clearly just a little bit of statistical or qualitative analysis alone isn't going to sway gcp to make changes to the pipeline.  

Since we seem to be at a place where we'll have plenty of time on our hands anyways, why not use it to prove our arguments and run actual training and test matches with our preferred modifications to see if they actually produce good results instead of just miring ourselves in theories?  Yes, it will be significantly slower than if it was tested with the project's resources, but why should that stop any of us?  Just get a few people who agree with you or are interested in your idea to help and spend a month or two to do it, then post your results here and schedule a test match against d018 with your network.  God knows it may very well still be the top network by then at this rate.

**Comment by l1t1** [comment]
@Crazyeight101 the question is the people suggest new method haven't enough resource, most of them could not prove their viewpoints in a peroid alone. and the project has much more resource.

**Comment by Crazyeight101** [comment]
@l1t1 which is why I suggest they garner support from their peers and do it as a group.  It may still be 20x slower than the project currently, but since they're not going to get the project's resources without convincing gcp anyways, it's not really worth weighing your options against.  Remember that gcp himself started this by asking for help from others and garnering support, and he initially only expected to get around 10-20 clients.  Isn't google also providing free access to decent GPU time for training?

**Comment by herazul** [comment]
The thing some don't seem to understand, and it's what's causing the confusion (for friday or marcin for example i think), is that the 55% gating IS NOT : 
"the new net have to have 55% WR versus the old net"
It is : 
"the new net have to be >50% wr, but we have to know this information in the fewer test matchs possible"

It is just because we can't have the %wr information with 100% confidence unless we do A SHIT TON of matchs, so there need to be a compromise between match number, %WR, and confidence.

The compromise chosen by deepmind is 400 games with 55% wr for a 95% confidence the new net is better than the old one, but it could be 52% with A LOT more test match.

If we lower the 55% gating to 52% for example, with the same number of test match, we are NOT saying : 
"we only need new net that are 52% stronger instead of 55% stronger than the old one"
we are saying : 
"we are ok with having a lot more net that are worse than the old one"

And we have have no way to know if it is better to trade some safety (with worse net doing the game generation) for some diversity. No one know for know what is best for the training, and no evidence.
We could try it sure, but no way to know if we wouldn't run into problems.
Maybe the net would still stall the same even if more 52-53% net pass because there would be a lot more false positive in these new nets, maybe it would even mess with the training and would hurt it.
Or maybe it would improve things. No one know.

**Comment by Friday9i** [comment]
@herazul: we should speak about that here #1524 : -). Hence, I'll do my answer there.

**Comment by gcp** [comment]
Increasing the amount of test games to say 800 or 1000 would allow lowering the gating to say 52.5% while keeping the exact same guarantees (I didn't do the exact math but this should be ballpark-correct)

But the reason I'm not terribly hot on lowering the gating threshold is that it simply expresses how much gain we can get from ~250k - 500k games of training data. If this drops to below 35 Elo, it means further progress is going to be *really* slow.

I think that's just a good sanity check that changes are in order. So far, it's been lowering learning rate, and increasing the network size. If we're at the lowest rate at 256x20 I personally think that's a sign we should look towards something else than 19x19 @ 7.5 komi that is more beneficial to Go players (0.5 komi network anyone?). I understand that some people will disagree there and want an even stronger 19x19 @ 7.5 and everything should be there to allow such a fork to work.

It is also true that eventually the growth curve would be expected to flatten off so that lower thresholds would be required. But reminder that this wasn't required for AGZ and not for ELF, so we should take that as a note that "something else is wrong" rather than just changing the threshold.

There's probably a theoretical argument to be made that some optimal gating exists that has a very minimal amount of false positives and allows marginally faster feedback of marginally stronger networks into the learning cycle, and that it likely isn't 55%. Is this particularly interesting for us? IMHO only in the theoretical sense. The fact that we're running with v=3200 right now is probably way, way worse (and I do intend to try to fix that when going to 256x20...).

**Comment by roy7** [comment]
Obligatory concerned comment that a much smaller visit count in training games might remove LZ's learned ability "read" ladders, vs ELF's "avoid" ladders.

**Comment by SebastienFaurite** [comment]
@roy7  are you sure ELF is always avoiding ladders ?
I mean is it possible for a network to see if it is interesting to read the ladder or simply ignore it ?
Like human do : if there is no stone on the other side of the board, you don't read it, you simply know it is good. But if there are multiple stones, you have to carefully read it completely.
So maybe ELF is just avoiding unnecessary reading. Have you seen an example where ELF should have read it or even taken the ladder, but did not ?

**Comment by Friday9i** [comment]
Thanks for your clarification @gcp!
I'm trying to make a synthesis of the optimal threshold approach (on #1524) but true, the impact is not mathematically provable and possibly not massive... It may still be nice, bit I understand your point.
And happy to know that v=3200 is under review, waiting a lot what you will propose ; -). Maybe a mix of v=800 to generate a lot of games quickly, and a proportion of deeper games (20% of v=3200, and a few % of v = 12800 games??). Anyway, let's wait and see
Thanks again for your hard work, looking forward for the changes!

**Comment by herazul** [comment]
i was thinking : if training a 10x128 on the current 15x192 games windows gives a way better 10x128 net than the one we trained then, we can conclude that to have the best possible net for our current personal hardware, a company with an insane amount of computing power could train an insanely good net of 160x1024 for example and then generate a big game windows with like 1 millions playouts with this net, and then train a 20x256 or a 40x256 net on this windows. The resulting net could be insane and work on current hardware.
We just need to rent a data center.
(this message is useless, was just a theory reflection)

**Comment by Friday9i** [comment]
@herazul: I proposed earlier something around that ;-). In fact, we could test it by training a 10x128 net (or even a smaller one) from current selfplay games, and see how it compares to the best 10x128 net to get the definitive answer! Unfortunately I'm a mathematician, not a programer, so not in a position to do it by myself...

**Comment by tapsika** [comment]
I think if LZ peaks at lower levels than other bots, that should be investigated and debugged.

For the various tweaks and options everybody randomly suggests, creating an experimental branch where 10-20% of the computing power is directed to seems like a good way to start. Then many of these ideas could be tried, and what works can be adopted back.

For the lower peaks: I'm not sure how much strength is lost by the defect of the policy loss, for example.

**Comment by nmcle** [comment]
I'd be very interested in a 0.5komi network that can play with handicap stones!
Much prefer that over a 40x256 block network, as the 15x192 is already very strong and a 20x256 would be strong enough.

I really like the LeelaMaster networks to play against too.

No idea how to train a 0.5komi handicap network, but it seems to be a good idea to use 2 networks of different strength and do selfplay games with varying handicap stones 2 to 9. Would be very nice to play against, please :) And then do a Master (human) version of it :)
Not sure about how test matches would work though...

Thanks agains to all the devs and clients with gpus! Thanks very much, I can play and review on a crappy laptop against a very strong dan level player.

**Comment by wonderingabout** [comment]
about 0.5 komi, is it not possible to play handicap games at 7.5 komi ?
since lz is already trained on 0.5 komi, i think making it understand and learn to play handi at a komi it knows may be actually easier ?
also, it would avoid the need to fork the 7.5 komi current branch

**Comment by afalturki** [comment]
@gcp How about experimenting with using [Global Pooled Properties](https://github.com/lightvector/GoNN#global-pooled-properties-dec-2017) or [Parametric ReLUs](https://github.com/lightvector/GoNN#parametric-relus-feb-2018)

@lightvector has shown they provide real benefits with no performance cost and GPP might be helpful toward achieving a multi-komi solution. Maybe a new 256x20 network can be trained with them and training continued while solutions for multiple komis are explored.

**Comment by Crazyeight101** [comment]
@nmcle A zero network trained to do handicap might be interesting and have broad appeal, but for the vast majority of players (<4d), Leela 11 should be a more than sufficient tool for handicap games and reviews.  On the other hand, for players like myself who play at or very close to a professional level, there's extreme interest and value in having an open source network trained to be as strong as possible.  There's the question of what is the true limit of the game?  What does the game, at levels well beyond current human understanding, look like?  For players like myself, this is really what's important, because we can then observe, study, experiment, find new interesting and important ideas about the game that humans currently lack, and pass those ideas on to the next generation.

LZ is certainly very strong now, but I'm still able to understand its moves.  What's more, I've learned how to exploit its weaknesses and can win against it reliably.  Compare this to AG master, which did show us some interesting new ideas but more so just the awesome power of computation in general, and to some of the AGZ 40b self-play games, which truly did surpass human understanding and left me and everyone else in the world just scratching our heads.

Honestly, when I looked at those AGZ 40b self-plays, that was the first time in a very long time that I could look at a game of go and not understand what the heck was going on, kind of like when I was a kyu player.  It was an incredible experience that filled me with both amazement and despair (considering how that iteration of AG got there in a little over a month, and over a 6 year span in total, whereas I had spent over a decade of studying just to nip at the heels of strong pros).

But we only have a handful of games from AGZ 40b to look at, and have no means to interact with it, so we can't really learn from it.  We're looking for LZ, or some other program, to fill this gap.  For the purposes you're seeking, there's already good tools out there, but for the purposes of strong players, no such tools exist yet.  Certainly you can appreciate how the novelty of an open-source superhuman network is far more valuable than having just another "strong enough" one.

**Comment by gcp** [comment]
@afalturki I was never able to get a consistent improvement for other activation functions when BatchNorm layers are in use. That is, you can get a miniscule improvement with ELU, PReLu, RPReLU etc but it's really a wash when you try it over longer terms. Without BatchNorm ELU works very well. I noticed ELF had PReLU code but disabled it. I personally would not spend any more time on this.

What would be more interesting to fix is to use full BatchNorm scale/center layers. Ttl was able to add center layers in our design by a hack of the weight format, and you can get both if you just take a network trained in one go (as with ELF conversion), but we can't use the scale layers in our training cycle. That they are missing is due to an oversight on my end on how BatchNorm and residual layers interact. (Scale layers are redundant if followed by a convolution, but NOT if they are followed by a residual add)

**Comment by gcp** [comment]
I would think that a 0.5 komi network can be trained with no changes to the current method (though I'd use v=800, T=1 and maybe start with 256x20 immediately). Of course the game outcomes will be lopsided but as long as the scores are not too extreme I would still expect it to work.

**Comment by gcp** [comment]
>LZ is certainly very strong now, but I'm still able to understand its moves. What's more, I've learned how to exploit its weaknesses and can win against it reliably

I assume you're talking about running it at very low visits, so the solution to get more strength and be closer to AGZ should be obvious here.

**Comment by alreadydone** [comment]
@gcp Wouldn't you add komi as an input feature as suggested in #1344? Then the network has a chance to generalize to other komi values as well. Plus we can augment training data by switching black/white and the sign of komi.

**Comment by gcp** [comment]
Yes, that would work. I wouldn't expect too much generalization aside from what is used in training though. But I guess you might do some mix of 0.5, 5.5, 6.5 and 7.5.

**Comment by gcp** [comment]
>We're looking for LZ, or some other program, to fill this gap.

If you look at the graph of A(G)Z training, there's a steep initial part and then a very, very long near flat-lined part.

Reproducing the "near flat-lined" part in a distributed project is not going to be a lot of fun.

**Comment by alreadydone** [comment]
Augmentation would provide training data for four values of komi even though we only produce games with two komi values. I imagine this would give the network enough hints for it to generalize.

**Comment by alreadydone** [comment]
Actually it seems reasonable to
+ first add the komi input feature
+ augment current training data as described
+ use the network after training on such data to generate 0.5 komi games

**Comment by Crazyeight101** [comment]
@gcp I'm playing against it running it on my pc hardware, and it's getting between 100k-500k visits per move, based on the type of position.  

The gap I was referring to isn't necessarily the gap in the strength between LZ and AGZ, it's the gap left by Deepmind by not giving us an engine to play against, despite showing us how strong it could be.  Maybe one day, in a few years or whatever time period it takes LZ and possibly subsequent projects, an open-source network will be produced that can match or beat AGZ 40b that we can freely play against on some high-end retail hardware and fill that gap.  

I mean, yes, it would be particularly nice if we could find some technique to shorten or to avoid having to deal with the long tail of the curve, and it's likely that a project like this one will choose to change its focus by the time we reach that point, but aren't we still a good while away from that, anyways?  By my estimation, we're still a fair ways away from even cracking AG Master yet, considering that it would win 100% of its games against the world's top pros within 30 moves or so on relatively minimal hardware.

nmcle had brought up a direction he would like to see LZ go in, so I just shared my view that there were already tools available that satisfied his desire (playing handicap games against and reviewing handicap games against a high dan player), as well as the direction I would like to see LZ go in.

**Comment by alreadydone** [comment]
It's a bit hard to measure progress on 0.5 komi though; in match settings black will almost always lose, dragging the overall winrate towards 50% even for networks that differ a lot in strength ... maybe some swap rule is in order. I am very curious what are balanced first moves under 0.5 komi, by the way.

@Crazyeight101 It would be great if you could compile a list of traps LZ would fall into, people like @bood and @Mardak are interested in tracking them.

**Comment by pcengine** [comment]
@Crazyeight101   
With all due respect, I have found it hard to believe that a human player can beat LZ RELIABLY with 100k-500k per move.  So my question is, which network did you play against?   

       Two months ago, there was an AI match between Leelazero and another very strong Chinese made AI (Golaxy). Before that match, that Chinese AI (Golaxy) gave one stone handicap (0.5 komi) to  30 different strong pro players, and has won 28 of 30.  Among those pro players, 4 have won world champions.   So maybe we can assume (with a 94% winrate giving 1 stone  against pros including 4 world champions) that AI(Golaxy) is much stronger than most pros?  But even that AI couldn't beat Leelazero RELIABLY.  Not to  mention that  Leelazero network was two months ago which means it's at least  a few hundred elo weaker than current best. So yes, it would be great if you can provide a few game files for people to learn.

**Comment by herazul** [comment]
@Crazyeight101 
Yeah same, i would love to see game where you managed to win against LZ, with the insane level of LZ, that must be very interesting.

**Comment by herazul** [comment]
@pcengine 
>>With all due respect, I have found it hard to believe that a human player can beat LZ RELIABLY with 100k-500k per move.

Yeah that's surprising, considering that with that amount of playouts LZ can win 2 stone handicap games against pros.
I would already consider it amazing to win an even game versus a 20k playouts LZ, and would love to see the games.

**Comment by lightvector** [comment]
Not anywhere close to a comparable number of playouts, but apparently Leela Zero still messes up ladders with 1000 playouts, badly enough that an amateur 4d-7d (depending on the online server) can still beat it sometimes in an even game. https://lifein19x19.com/forum/viewtopic.php?p=232587#p232587

**Comment by herazul** [comment]
@lightvector : Yeah but it's a specific case, LZ is already very strong at 1000 playout, unless you make it fail a specific situation like ladder, and it's not EZ at all.
But these mistakes wont happen with a reasonable amount of playouts, i don't even know if you can make it fail a ladder at 10k or 20k playouts.
1000 playouts is nothing, i can almost have that on my smartphone...

**Comment by Friday9i** [comment]
I made 4 games of LZ148 against Elf (3200 visits for ELF), while giving it 2 stones! And ...:
- 100K and 200K visits were not enough for LZ, it almost worked as it went up to ~45% winrate in the 200K game, but it collapsed in the end
- 400K and 1M (!) visits were enough, it managed to beat ELF while giving it 2 stones!
Knowing that ELF is pro level at 3200 visits (around 3500 Elo on CGOS), I'm also surprised that a recent LZ is beatable by a pro without handicap... 
Note: I posted the 1M game here https://www.reddit.com/r/cbaduk/comments/8s1s9e/for_fun_lz148_crushes_elf_with_3200_visits_while/
BTW, if someone knows if it is possible to save the GTP console info in Sabaki, please tell me how :-).

**Comment by FeoDoT90** [comment]
I think the size of the nets is not important, but the quality of the food provided for the net is important. For every million game samples, it will take another two million games to train, so that the network can significantly Refine its coefficients. In my opinion, this is logical, because here the key role is played by science - statistics. Therefore it is impossible to adopt the accumulated experience of old networks, for networks of new and large sizes. Otherwise, we rely only on luck that one day the network will appreciate the once rejected course of action from numerous weak experiences. After all, the essence of training is not to feed the network a lot of games generated exclusively by accident, or human experience, or a competitor's game, but to give it a little more powerful experience that the network can correctly learn. If we continue to invest in the existing experience of existing networks, we will grow a very short-sighted network, and I think we will have to make much more effort than it takes to reach level of AGZ. Just rate the growth graph of the force LZ, he is not like the progress chart of AGZ.

My moods can be very depressing and radical, but I urge you to give up the old network experience, and start from scratch to reach higher heights in a few months, with the final size of the 40x256 network.

I think we need to forget about the artificial ELF network, the origin of which we know nothing about, and stop chasing other competitors. We have already done a miracle worthy of admiration, but I think it is necessary to settle down before it's too late, to abandon the high-speed race, and start our flight of strength growth with absolute chaos of zero experience.

**Comment by kityanhem** [comment]
@pcengine 
Sr but on KGS even some 8d japanese beat petgo3 4-0.

**Comment by alreadydone** [comment]
@gcp Hmm, embarassingly it took me one day to realize what "the network inputs are side-to-move already" really means: it means that
+ when the black-to-move plane is 1 and white-to-move plane is 0, the current player assumes 7.5 komi;
+ when the black-to-move plane is 0 and white-to-move plane is 1, the current player assumes -7.5 komi.

But then maybe we can just use the color planes as the komi plane! 
I propose to fill the black-to-move plane with 0.533=(0.5+7.5)/(7.5+7.5) and the white-to-move plane with 0.467 when black is to move, and 0.467 and 0.533 when white is to move, and see whether this makes the network play reasonably under 0.5 komi. If so, we would have a nice initialization for generating 0.5 komi games!

**Comment by Eddh** [comment]
I think it's only a few thousands playouts per move for petgo3

**Comment by jkiliani** [comment]
Using the color planes as komi planes sounds interesting, but the exact way to normalise those planes would have to be planned carefully. Leela Chess had problems resulting from an input plane that wasn't normalised (the 50-move rule plane).

**Comment by alreadydone** [comment]
We can say currently it's normalized so that 0 ~ 1 corresponds to -7.5 ~ 7.5 komi. Larger komi would be extrapolation, but maybe useful for better handicap performance. It's usually reasonable to expect that interpolation works well IIRC.

**Comment by pcengine** [comment]
@kityanhem 
> Sr but on KGS even some 8d japanese beat petgo3 4-0.
LZ still has some big weakness like semeai and seki.

Are you aware the strength differences between a few thousands playout and 100k-500k playouts per move?  As many have pointed out, it's very easy to settle the issue by providing some sgf files. Many people would be interested to see those **BIG WEAKNESS** at 100k-500k playouts per move. Since someone has claimed that he can RELIABLY beat LZ with 100k-500k playouts per move, maybe uploading such files should not be as difficult?

**Comment by herazul** [comment]
@kityanhem there was a thread on petgo3 : https://github.com/gcp/leela-zero/issues/1501
this seems to be kinda specific to petgo3, i don't think usual LZ well configured has any big weakness, or i would like to see them too (excluding ladder or specific situations with very low playouts)

**Comment by odeint** [comment]
Petgo3 is LZ on a GTX1060 so people can still beat it using LZ on a 1080, or ELF.

**Comment by Hersmunch** [comment]
There’s an couple of comments about interpolating for different Komi here https://github.com/gcp/leela-zero/issues/1221#issuecomment-382535040

**Comment by Friday9i** [comment]
My computer currently plays a game of LZ150 @ 500K visits against ELF @ 1000 visits, giving it 3 stones!!
And... LZ150 may well win the game as it is now coming close to a 50% win rate (ELF still at 70%+ win rate) in mid-game. Result tonight (with the game & detailed files of moves, depth of analysis, etc...).
Knowing ELF @ 1000 visits is about pro level, it's another hint to show that LZ@500K visits is very clearly superhuman... Hence, winning against it regularly seems surprising.

Thinking about it and looking for a good faith explanation, 2 ideas:
- It may be an older version of LZ: if it is around LZ100/LZ120, that seems credible. 
- Maybe @Crazyeight101 has a slow computer AND he did not specify "time_settings 0 1 0". I fell for that trap earlier, playing some games with 500K visits or more but, in fact, it was time-constrained and played within 30 seconds and with "only about 30K visits". With a slow computer (without GPU), 30s is much less than that, a few 100 visits: that may explain he specified 500K visits in good faith and beat it, but in reality it was only 100s visits (that would still be quite an achievement by the way ;-).

@Crazyeight101, would you mind to check that and clarify the situation?

**Comment by alreadydone** [comment]
So the color plane interpolation idea is not new and mentioned here: https://github.com/Chicoryn/dream-go/issues/25#issuecomment-381417656
The list of input features look attractive and one may explore whether adding them improves small net strength when trained on recent data.

@Crazyeight101 did say 100k-500k per move based on the type of position.

**Comment by pcengine** [comment]
@Friday9i 
Thank you for providing emprical evidence explaining how unlikely for a human player  to beat current LZ.

> Thinking about it and looking for a good faith explanation, 2 ideas:
It may be an older version of LZ: if it is around LZ100/LZ120, that seems credible.

Well, if your read his comments, you can easily  see that  he is quite eager to learn newest and strongest moves possible. Hence, it is very difficult for me to imagine that such type of player would just  sit there not updating any newer and stronger weights for more than 2 months. What do you think?

**Comment by Friday9i** [comment]
@pcengine: yeah, I share your view. @alreadydone: same, I saw that part, but maybe he played 100K games and 500K games and he made a shortcut, not impossible.
All in all, I'm sceptical, but I still suppose that he said what he said in good faith. If we don't get additional info from him, then it may not be entirely true...

**Comment by hydrogenpi** [comment]
@Crazyeight101

Each of the gen 2 TPU's consists of 180 TeraFLOPS for a total of 720 TeraFLOPS...
The matches via Haylee that were on 4x of Nvidia V100 used OpenCL, so they did not even take advantage of its Tensor deep learning modules. So roughly what Leela Zero had to work with was actually merely about 5 TeraFLOPS per V100 for a total of 20 TeraFLOPS... maxed out. 

So in the Master/AlphaGoZero setup and games against Lee Sedol, Ke Jie, 60 top pros, etc etc Deepmind used 36 times more processing power than Leela Zero ever did during its exhibition games with Haylee and even the unofficial games that Baduk1 played against some of the top ten Go players (and which Leela Zero convincingly won on even without any handicaps either sides)  [note in the match against Lee Sedol, Deepmind used 48 TPUs instead of just 4 TPUs]

So we are talking about more than five doublings in power between the hardware Deepmind used and the hardware Leela Zero has ever maximally used. At the lower end of 80 elo per doubling and higher end of 120 elo per doubling, this amounts to between a 400 elo to 600 elo raw advantage that AlphaGo would have against leela zero, merely on just hardware configuraton differences alone. 

So your statement of : "" By my estimation, we're still a fair ways away from even cracking AG Master yet, considering that it would win 100% of its games against the world's top pros within 30 moves or so on relatively minimal hardware."" is itself understating the power that Deepmind had access to and had indeed utilized in its matches/games and trivializing the role that horsepower had to do in all of this to max the AlphaGo shine as brightly as it did. 

There are difference levels of superhuman just like there are different levels of supersonic. The Bell X-1 and the Space Shuttle are both supersonic, but one is about 25 times faster than the other. While there is no doubt Leela Zero still has a long ways to go before ever catching up to Alpha Go Zero's level, it has already surpassed AlphaGoLee when on energy parity and when using hardware consistent with the lower bounds of what Deepmind/Google used, there is no doubt Leela Zero is already superhuman. 

Therefore I find your claims of being able to consistently and/or reliably beat Leela Zero (on even I assume) at very high playouts/visits to be dubious at best. Perhaps like others have remarked it might be a good faith innocent mistake of game configurating settings on your part... However if that is not the case, then surely if you can achieve something as remarkable as finding and exploiting weak spots of current Leela Zero to defeat it reliably then supporting those claims with publishing of sgf and logs should not be too much of a problem either.

**Comment by bjiyxo** [comment]
@hydrogenpi You man try to analyze Master 55 games with LZ 2m playouts, then you will find we are really far behind Master. I already did that, and found out lz still has lots more to work on.

**Comment by Friday9i** [comment]
I noticed that too: LZ has no clue of the position in top AG or AZ games, it does not manage to get a clear view of who's winning/losing, and IIRC it's sometimes not even clear when 1 player resigns...
So LZ is already superhuman, but it is still (very) far from top AG/AGZ/AZ level! That's just amazing

**Comment by Eddh** [comment]
We can expect ELF to be close to AGZ 20 block strength, but AGZ 20 block is hundreds of elo weaker than AG master or AGZ 40 block so it's not that surprising.

**Comment by alreadydone** [comment]
https://deepmind.com/blog/alphagos-next-move/ "... a special set of 50 AlphaGo vs AlphaGo games, played at full length time controls."

From AGZ paper:

"In our evaluation, all programs were allowed 5 s of thinking time per move; AlphaGo Zero and AlphaGo Master each played on a single machine with 4 TPUs; AlphaGo Fan and AlphaGo Lee were distributed over 176 GPUs and 48 TPUs, respectively. 

The raw neural network, without using any lookahead, achieved an Elo rating of 3,055. AlphaGo Zero achieved a rating of 5,185, compared to 4,858 for AlphaGo Master, 3,739 for AlphaGo Lee and 3,144 for AlphaGo Fan.

Finally, we evaluated AlphaGo Zero head to head against AlphaGo Master in a 100-game match with 2-h time controls. AlphaGo Zero won by 89 games to 11."

So most likely the 50 Master games are played under 2h time control on 4 TPUs.

**Comment by bjiyxo** [comment]
On Aja Huang's Facebook,

> 50 AlphaGo vs AlphaGo games (AlphaGo's thinking time was 2 mins for every move), which we believe contain many new and interesting ideas and strategies.

So its 2 mins per move.

**Comment by alreadydone** [comment]
"Color planes for komi" implemented, see https://github.com/gcp/leela-zero/issues/1576.

**Comment by alreadydone** [comment]
"Dynamical komi" to keep winrate within a certain range now implemented in https://github.com/gcp/leela-zero/issues/1599.
